modules/imgcodecs/src/loadsave.cpp

```diff
@@ -492,6 +532,136 @@ imread_( const String& filename, int flags, Mat& mat )
     return true;
 }
 
+/**
+ * Read an image into memory and return the information
+ *
+ * @param[in] filename File to load
+ * @param[in] OutputArray Reference to C++ Mat/Umat/GpuMat object
+ * @param[in] ImreadParams flags, maxPixels, maxSize options
+ *
+*/
+static ImreadError
+imread_2( String const& filename, OutputArray image, ImreadParams params)
+{
+    /// Search for the relevant decoder to handle the imagery
+    ImageDecoder decoder;
+    ImreadError error = IMREAD_OK;
+    int flags = params.flags;
+
+#ifdef HAVE_GDAL
+    if(flags != IMREAD_UNCHANGED && (flags & IMREAD_LOAD_GDAL) == IMREAD_LOAD_GDAL ){
+        decoder = GdalDecoder().newDecoder();
+    }else{
+#endif
+        decoder = findDecoder( filename, error );
+#ifdef HAVE_GDAL
+    }
+#endif
+
+    /// if no decoder was found, return nothing.
+    if( !decoder && error != IMREAD_OK){
+        return error;
+    }
+
+    int scale_denom = 1;
+    if( flags > IMREAD_LOAD_GDAL )
+    {
+        if( flags & IMREAD_REDUCED_GRAYSCALE_2 )
+            scale_denom = 2;
+        else if( flags & IMREAD_REDUCED_GRAYSCALE_4 )
+            scale_denom = 4;
+        else if( flags & IMREAD_REDUCED_GRAYSCALE_8 )
+            scale_denom = 8;
```

Comment:
Semantically this should be a direct field of `ImreadParams`.
Otherwise there is unresolved mess between `flags` and `ImreadParams` in designed API.

Reply:
@alalek Do you mean that i should split flags into scale_denom and loadGdal variable and put it inside ImreadParams?